### ISSUE TITLE
[3.x] GDScript: Bugs with preloaded constant shadowing native class

### ISSUE THREAD

**Issue by schweigert** [godotengine/godot#57496]
### Godot version

3.3.3

### System information

Windows 11, Windows 10, Ubuntu, Manjaro and MacOS

### Issue description

When you type a function called by a signal, the variable receive `null`.

### Steps to reproduce

```gdscript
# OtherScript.gd
extends Node

signal call(obj)

func _ready() -> void: emit_signal('call', self)
```

```gdscript
# Script.gd
extends Node

const ButtonGroup = preload('res://OtherScript.gd')

func on_call(obj: ButtonGroup) -> void: print(obj)
```

Connect signal `OtherScript.gd@call` to `Script.gd@on_call`

Output:

```
null
```

Remove the typing from `Script.gd`:

```

# Script.gd
extends Node

func on_call(obj) -> void: print(obj)
```

Now, the output shows the object.

**Comment by dalexeev** [comment]
I have tested this in 3.5.2 and 4.1.

The reason for the bug in 3.x is shadowing a native class. You are using the identifier `ButtonGroup`, which is the same as the native class [`ButtonGroup`](https://docs.godotengine.org/en/3.5/classes/class_buttongroup.html). The following code also causes the bug:

```gdscript
extends Node

const ButtonGroup = preload("res://test.gd")

func _ready() -> void:
    call("test", ButtonGroup.new())

func test(value: ButtonGroup) -> void:
    print(value)
```

The error is that `ButtonGroup.new()` creates an instance of the custom class, but type hints still treat it as the native type. (In your example, you are passing `self`, which is also an instance of the custom class.)

```gdscript
print(ButtonGroup.new().get_class()) # Node
```

The following examples expose this error:

```gdscript
# Error: The passed argument's type (test.gd) doesn't match
# the function's expected argument type (ButtonGroup).
test(ButtonGroup.new())
```

```gdscript
# Error: The assigned value type (test.gd) doesn't match the variable's type (ButtonGroup).
var t: ButtonGroup = ButtonGroup.new()
```

But when the function is called from outside, via `call()` or via a signal, this error can only show up at runtime. Removing the type hint "fixes" this bug because it not only removes the static type, but also changes the `GDScriptDataType` that is checked at runtime.

Note that this error only happens with `const` - `preload`. If you try to declare a shadowing nested class, you will get an error:

```gdscript
# Error: The class "ButtonGroup" shadows a native class.
class ButtonGroup:
    pass
```

There is no bug in 4.x because GDScript doesn't allow shadowing a native class even with `const` - `preload`:

```gdscript
# The member "ButtonGroup" shadows a native class.
const ButtonGroup = preload("res://test.gd")
```

I think we can add the same error message in 3.x. Making shadowing work properly is not trivial and probably too risky for the LTS branch.

**Comment by akien-mga** [comment]
Fixed by #79123.
